### PR TITLE
BatchDecoupledVariationalStrategy

### PR DESCRIPTION
**Pull request by gpleiss**
Implements MFD models from Parametric Gaussian Process Regressors.

Docs look like this:
Screen Shot 2020-06-12 at 9 35 37 AM

### CODE REVIEW

File: gpytorch/variational/batch_decoupled_variational_strategy.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+        # Compute interpolation terms
+        # K_ZZ^{-1/2} K_ZX
+        # K_ZZ^{-1/2} \mu_Z
+        L = self._cholesky_factor(induc_induc_covar)
+        if L.shape != induc_induc_covar.shape:
+            # Aggressive caching can cause nasty shape incompatibilies when evaluating with different batch shapes
+            del self._memoize_cache["cholesky_factor"]
+            L = self._cholesky_factor(induc_induc_covar)
+        interp_term = torch.triangular_solve(induc_data_covar.double(), L, upper=False)[0].to(full_inputs.dtype)
```
[the remark this answers is not in the document]
Comment (gpleiss) [reply gpytorch/variational/batch_decoupled_variational_strategy.py:164]: Yeah we use it in the default variational strategy.

File: test/variational/test_batch_decoupled_variational_strategy.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        return gpytorch.variational.MeanFieldVariationalDistribution
+
+
+class TestMeanFieldBatchDecoupledRobustVGP(TestBatchDecoupledRobustVGP):
+    @property
+    def distribution_cls(self):
+        return gpytorch.variational.MeanFieldVariationalDistribution
+
+
+class TestBatchDecoupledVariationalGPBatchDim(TestBatchDecoupledVariationalGP, unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (gpleiss) [reply test/variational/test_batch_decoupled_variational_strategy.py:89]: no... i can fix that

**Comment by jacobrgardner** [comment]
Test failure in both this as #1180 seem unrelated to the PRs, looks like the SM kernel test has just become flaky

File: gpytorch/variational/batch_decoupled_variational_strategy.py [hunk 3]
```diff
@@ -12,9 +12,9 @@
 class BatchDecoupledVariationalStrategy(VariationalStrategy):
     r"""
     A VariationalStrategy that uses a different set of inducing points for the
-    variational mean and variational covar.  It is roughly based on the model
-    proposed by `Cheng et al. (2017)`_, but with a few differences (proposed by
-    `Jankowiak et al. (2020)`_).
+    variational mean and variational covar.  It follows the "decoupled" model
+    proposed by `Jankowiak et al. (2020)`_ (which is roughly based on the strategies
+    prooposed by `Cheng et al. (2017)`_.
```
Comment (martinjankowiak) [inline gpytorch/variational/batch_decoupled_variational_strategy.py:17]: typo

File: test/variational/test_batch_decoupled_variational_strategy.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gpleiss) [reply test/variational/test_batch_decoupled_variational_strategy.py:89]: added a test

**Comment by gpleiss** [comment]
@jacobrgardner and @martinjankowiak - to be consistent with #1180 - the `mean_var_batch_dim` and `latent_dim` attributes require negative indices, where `-1` corresponds to the right-most batch dimension.

(Before in this PR - `-3` referred to the right-most batch dimension. I updated the module to be consistent with #1180)

**Comment by jacobrgardner** [comment]
@gpleiss I'm not sure how I feel about relative indexes only indexing in to the batch shape -- this doesn't occur elsewhere in the package AFAIK. On the other hand, I'm not sure we expect end users to ever index in to a batch dimension like this elsewhere in the package, so maybe this sets a precedent rather than breaks one? Thoughts?

**Comment by gpleiss** [comment]
@jacobrgardner - I'd be fine with having it either way (-1 or -3), but -1 more closely matches what Pyro does with batch dimensions. Indexing by batch dimension I think makes more sense when the objects may have different "event" dimensions (e.g. the first batch index is -2 for GP outputs, but -3 for GP inputs).

I also don't think we've indexed batch dimensions from GP models yet (just from LazyTensors), so this would be setting a precedent.

**Comment by jacobrgardner** [comment]
> @jacobrgardner - I'd be fine with having it either way (-1 or -3), but -1 more closely matches what Pyro does with batch dimensions.

I guess part of the reason this makes sense in pyro is they more explicitly delineate batch shape / event shape than we do (I know LazyTensor technically has `batch_shape` and `matrix_shape`).

I guess we can go with this for now and revisit it if it becomes a source of confusion.

**Pull request merged**